### DISCUSSION TITLE
Acid Hydras anyone?

### DISCUSSION THREAD

**Discussion by a1exsh** [ihhub/fheroes2#4524] [Show and tell]
Hi!

As I mentioned earlier, I was exploring what it takes to add new monsters to the game (per my son's request) and here's a sneak peek of an upgraded Hydra unit, Acid Hydra :)

https://user-images.githubusercontent.com/489601/139539675-efcf9241-bc6a-44b1-bb19-6e65cb0d80d3.mp4

The changes to the code I've made to get this working can be found on this branch: https://github.com/ihhub/fheroes2/compare/master...a1exsh:hydra2

When it comes to the images, we've simply extracted the PNG files using `icn2img` from `HYDRA.ICN` and changed the color hue using Gimp's filters (which turned out to be scriptable as well).  This method is particularly promising, IMO, as it allows to add new upgraded units programmatically, without the need to distribute copyrighted resources from the original game: the same algorithm as Gimp uses can be implemented in the code if needed.

The images are then loaded from a directory under `~/.local/share/fheroes2/icn/`, if they exist, otherwise standard game resources are queried.  I didn't dare adding an XML parser, so `icn2img` was also changed to output a simple text file with the sprite offsets, and the PNGs are saved with real transparency instead of the gray background.

That's all, folks! ;)

PS: I do not intend to open a PR with these changes — I will keep them in that private branch.  Would love to hear about plans for extensibility, though: what others have explored so far, what is feasible to extend with a scripting language instead of hard-code, etc.

**Comment by a1exsh** [comment]
Just discovered that the PNG loader has R and B channels mixed up.  Here's a fix: https://github.com/ihhub/fheroes2/commit/f7a96104c86345efc4abcc2f77d5a29ff57d5948

With our hydra that's barely noticeable because it's mostly green and gray colors. ;)

**Comment by LeHerosInconnu** [comment]
Hello @a1exsh,

For your information, the video in the first post is not viewable.

I think that recoloring already existing sprites is okay for personal use.
But frankly, it's not really my cup of tea.
I really wouldn't want to be faced with yellow Goblins, red Goblins, blue Goblins, orange Goblins, purple Goblins, and I saved the best for last, pink Goblins. :D

**Reply by a1exsh** [reply]
> Hello @a1exsh,
> 
> For your information, the video in the first post is not viewable.

That's a pity, esp. given how much effort it taken me to make GitHub accept it. It plays OK on my laptop and smartphone though...

> I think that recoloring already existing sprites is okay for personal use. But frankly, it's not really my cup of tea. I really wouldn't want to be faced with yellow Goblins, red Goblins, blue Goblins, orange Goblins, purple Goblins, and I saved the best for last, pink Goblins. :D

:D

**Reply by LeHerosInconnu** [reply]
It is possible to view it in VLC by downloading it, but I can't play it with either Firefox or Chrome.

**Reply by zenseii** [reply]
I'm able to view the video just fine from my PC using chrome.

**Reply by Branikolog** [reply]
Agree with @LeHerosInconnu, changing of colors looks like "cheap" improvement, which is acceptable only for a few exceptional cases, when a creature has well-known other form with a different material/color/nature. For example mages: Black/White; dragons: Red/Green/Gold; or golems: Steel/Wood/Aluminum.
OG has some similar creatures and making more "color upgrades" looks not so professional... From my point of view...
Nevertheless, I still expect at least some tiny visual differences even in the original game creatures for these upgrades. 
On subject: I'd like Hydras have some differences in sprite, not only color. Like second tail, for example... Or some kind of fashionable hats for pink Goblins.

Btw, video in the first message is visible to me but seems to have some problems with playback speed.

**Reply by a1exsh** [reply]
> changing of colors looks like "cheap" improvement

Sure. At the same time, almost all of the upgraded creatures in OG are just recolored original sprites, with the notable exception of the Knight castle.

So this approach is pretty much in line with the OG, IMO. It is also feasible to implement without involving a professional artist. ;)

**Reply by Branikolog** [reply]
Cannot say, "almost all" here... Orcs, Trolls, Elves, Dwarves, Liches, Swordsmen, Pikemen, Cavalry, Archers have differences in sprites. Actually, I've just checked: the major differences (If we won't mention colors) are hats. :D (Note, that not all creatures have upgrades at all, so this is not the small part of "professionally improved" creatures. ;)

**Reply by ihhub** [reply]
@a1exsh and @Branikolog , yes, most of upgraded creatures in the game are recoloured with minor changes. Others have some notable changes upon upgrades.
We need to be extra careful while adding new non-neutral creatures in the game by two reasons:
- fans expect new good quality content so in their eyes recolouring might not look great (I'm not saying that your solution is bad). This is what @Branikolog and @LeHerosInconnu are trying to say. In this case visual difference must be not only in colour but in some attributes of the creature (like pink Goblin with flowers 😅)
- Balance. This is the biggest issue. All new creatures must not break game balance. Warlock castle is too overpowered and adding even a more powerful Hydras might break the balance completely. The same applies for any monster in the game being added.

I'd say this is good start towards adding a new content to the game but it requires little more polishing.

**Reply by a1exsh** [reply]
Fair points — I agree with all that. I don't myself think Warlock needs an upgraded hydra urgently (unless other factions have ramped up their troops themselves) — that just was my kid's request ;)

I want to share the outcome of this experiment so it doesn't get lost and maybe we learn something for future work.

**Reply by ihhub** [reply]
Absolutely. As I wrote it's a good start and as I can see we need to do some architecture redesign to allow new creatures to be added in an easier way.

**Reply by LeHerosInconnu** [reply]
> (like pink Goblin with flowers :sweat_smile:)

And especially if he is going to San Francisco. :rofl:
https://www.youtube.com/watch?v=LlVwcBTXwmo

**Comment by a1exsh** [comment]
> It is possible to view it in VLC by downloading it, but I can't play it with either Firefox or Chrome.

@LeHerosInconnu can you try with the following one (used a different video codec):

https://user-images.githubusercontent.com/489601/141255684-0deb3acf-59aa-4914-9a27-39b5e6ea89d2.mp4

**Reply by LeHerosInconnu** [reply]
Hello @a1exsh,

> > It is possible to view it in VLC by downloading it, but I can't play it with either Firefox or Chrome.
> 
> @LeHerosInconnu can you try with the following one (used a different video codec):

It's the same, with a lower video quality. :)
I suggest you to try with OBS Studio.

**Reply by zenseii** [reply]
I'm not able to see this other video.

**Comment by a1exsh** [comment]
Out of curiosity I've created composite images of base and upgraded creature sprites to highlight the re-coloring and details (some of this I have never noticed):

- Knight
   - ![image](https://user-images.githubusercontent.com/489601/141612644-3ed1a698-40ec-43b4-9060-bec6e9ccc83b.png) Ranger got a fancy hat with blue feathers, the rest is identical but colored differently
   - ![image](https://user-images.githubusercontent.com/489601/141612679-c70069d4-39e7-4ce4-ae37-1a4363453c44.png) Veteran Pikeman got blue feathers on his hat
   - ![image](https://user-images.githubusercontent.com/489601/141612690-42ab491c-f119-41b5-ad53-52b62b6afcca.png) Master Swordsman got blue feathers and blue stripes on the shield
   - ![image](https://user-images.githubusercontent.com/489601/141612709-28c8694a-235a-440a-8333-58bfe227c84d.png) Champion: bigger, blue feathers (and some for his horse as well ;)
   - ![image](https://user-images.githubusercontent.com/489601/141612735-0201dd9b-d803-4e48-8bdf-14feb390acc7.png) Crusader: you guessed it, blue feathers instead of red ones
- Sorceress
   - ![image](https://user-images.githubusercontent.com/489601/141612766-a1e52b35-eef9-4892-acbf-bdacecf5d9d8.png) Battle Dwarves have longer horns on their helmets!
   - ![image](https://user-images.githubusercontent.com/489601/141645000-ef3f1ef8-c8f9-446a-9ffd-afb3150d28f2.png) Grand Elves: bigger blue feathers and a chain mail (never noticed that one!)
   - ![image](https://user-images.githubusercontent.com/489601/141612781-83818e71-d62c-45dc-9d23-54c7d3a8c1dd.png) **Here comes the interesting one.** The Greater Druids are exactly of the same shape as the good ol' Druids, but recolored.  Look closely at the chin and left hand and you'll spot a few pixels that apparently got re-colored automatically from the brown hue to the new fancy purple, but they shouldn't have been!
- Barbarian
   - ![image](https://user-images.githubusercontent.com/489601/141612860-489bdcd3-e96b-4412-9c0a-e0b8067b3c40.png) Orc Chiefs got a bigger hat with a spike on it
   - ![image](https://user-images.githubusercontent.com/489601/141613090-218ef195-551f-45e7-aa38-aae601c886e9.png) War Trolls also got some golden thing with a red jewel to hold their hair properly (have never noticed it before) ;)
   - ![image](https://user-images.githubusercontent.com/489601/141612895-e2a0b9a8-43b7-4848-a2d1-26e6beae4274.png) Ogre Lords are only re-colored as far as I can tell
- Warlock
   - ![image](https://user-images.githubusercontent.com/489601/141612907-de1c76fe-1e5d-47a9-ba13-e91d2d20b029.png) Minotaur Kings are just re-colored
   - ![image](https://user-images.githubusercontent.com/489601/141612918-f13adde9-b5b7-4c56-953b-ec44f5bc5a28.png) All three breeds of Dragons from the Warlock castle are the same shape just colored differently (which is a pity — more differences would be appreciated from OG)
- Necromancer
   - ![image](https://user-images.githubusercontent.com/489601/141612946-1c50ff12-765d-471e-a831-ba086c2673d5.png) The Mutant Zombies lost all hair instead of getting some and their head seems to have nodded a bit compared to the normal Zombies
   - ![image](https://user-images.githubusercontent.com/489601/141612977-14a7c869-2381-4326-864a-143fb8e9d38b.png) The Royal Mummies are just the normal ones, recolored
   - ![image](https://user-images.githubusercontent.com/489601/141612987-2a57ac0f-19d7-41e1-a062-b26869cd64ec.png) The same for Lord Vampires
   - ![image](https://user-images.githubusercontent.com/489601/141613001-b6e8267e-31ba-46d7-8b0b-e5eed7347565.png) Power Liches at first glance differ substantially from the normal Liches, but that's again mostly recoloring, but they did get a bigger crown ;)
- Wizard (at last!)
   - ![image](https://user-images.githubusercontent.com/489601/141613023-dd044f92-8fe6-4b27-9033-b9c282ceb69c.png) Steel and Iron Golems are no mystery: only recolored
   - ![image](https://user-images.githubusercontent.com/489601/141613040-194536b5-257c-46d2-8f04-a5cece987024.png) Archmages also got no extra detail compared to Mages, sadly
   - ![image](https://user-images.githubusercontent.com/489601/141613062-5a6d26aa-6919-4c65-b2ef-ce2c27e5ad87.png) And finally, Titans are just Giants colored more diversely

That's all, folks! :D

Upd: forgot about Elves, added.

**Reply by LeHerosInconnu** [reply]
@a1exsh,

> Out of curiosity I've created composite images of base and upgraded creature sprites to highlight the re-coloring and details (some of this I have never noticed):

Note that in Heroes 2 only human-like creatures have an upgrade. :)
The only exceptions are level 1 creatures, and Dragons.
The Dragon also has an exception of having two upgrades.

**Reply by ihhub** [reply]
@a1exsh , amazing findings! We might need to fix Great Druids colouring issues.

I fully agree that all upgrades look mostly like recoloured but if we do the same way many fans will go into rage mode and throw into us rotten tomatoes for so-called "cheap" work. This is due to a fact that players are used to existing monsters in the game and they expect something "new", not just recolouring.

**Reply by a1exsh** [reply]
> Note that in Heroes 2 only human-like creatures have an upgrade

Never thought about it this way, but seems to be true. Then the monster who is actually lacking an upgrade is Cyclops! :D

> We might need to fix Great Druids colouring issues.

Well, one can only see these after zooming in heavily, so I wouldn't rush it. ;)

**Reply by LeHerosInconnu** [reply]
Hello @a1exsh,

> > Note that in Heroes 2 only human-like creatures have an upgrade
> 
> Never thought about it this way, but seems to be true. Then the monster who is actually lacking an upgrade is Cyclops! :D

And Nomad, Ghost, Genie, Medusa and all Elementals, and Bone Dragon is also a Dragon and then can be another exception.
Just because it's possible, doesn't mean it has to be done. :D

**Comment by Schlumpf** [comment]
If you are looking for new content, take a look at the HoMM3 mod "The Succession Wars": https://heroes2.forumactif.com/f24-development-board
They have 2 new fractions with new creatures.

**Reply by LeHerosInconnu** [reply]
Hello @Schlumpf,

> If you are looking for new content, take a look at the HoMM3 mod "The Succession Wars": https://heroes2.forumactif.com/f24-development-board They have 2 new fractions with new creatures.

While this may eventually be feasible for personal use, we will not "plunder" the work done by the designers of another mod (or even use it with their consent).

Fheroes2 must have and keep its own character. :)

**Reply by Schlumpf** [reply]
Don't steal their work!
That's not what I was suggesting. Perhaps see it as a source of inspiration. They have also done some research on balancing.

**Reply by a1exsh** [reply]
> Don't steal their work!
> That's not what I was suggesting. Perhaps see it as a source of inspiration.

Hm, while I have my reservations towards their implementation of the Fortress castle, I do find the new creatures in Dervish castle pretty decent: the upgraded Rogue (what was the name — Bandit?) is exactly the type we are looking for here, re-colored and more details added.  The completely new creatures from the same faction: Acolyte, Anubian and Manticore(?) are also following the style of OG really well, IMO.

What if we try to contact the artist(s) who did this work for that mod and ask if they'd be interested to contribute to this project? :D

**Reply by LeHerosInconnu** [reply]
Hello @Schlumpf and @a1exsh,

> Don't steal their work!
> That's not what I was suggesting. Perhaps see it as a source of inspiration. They have also done some research on balancing.

Seen in this sense is okay, I think. :)
Regarding balance, I like the way the game is balanced in Heroes 2, depending not only on castles and creatures, but also on the size of the map.
And I like playing all castles on all map sizes. :)

> What if we try to contact the artist(s) who did this work for that mod and ask if they'd be interested to contribute to this project? :D

:)

**Comment by Tromthar** [comment]
+1 to this idea, im into some optional modding to this game, thanks for ur ideas alex.

**Comment by a1exsh** [comment]
Just as a means of backup, attaching the script I've used for the Hydras. ;)
```py
#!/usr/bin/env python3
import os
import subprocess

subprocess.run(
    ["gimp", "-i", "-c"] +
    sum(
        [
            [
                '-b',
                f"""
(let* ((img (car (file-png-load RUN-NONINTERACTIVE "{f}" "")))
       (drw (car (gimp-image-get-active-drawable img))))
  (gimp-drawable-hue-saturation drw HUE-RANGE-YELLOW 180 -100.0 0 0.0)
  (gimp-drawable-hue-saturation drw HUE-RANGE-RED 180 0 0 0.0)
  (file-png-save-defaults RUN-NONINTERACTIVE img drw "../../icn/HYDRA3.ICN/{f}" "")
  (gimp-image-delete img))
"""
            ]
            for f in sorted(os.listdir())
            if f.endswith(".png")
        ],
        []
    ) +
    ["-b", "(gimp-quit 0)"]
)
```
